modules/miner/miner.go

```diff
@@ -194,18 +210,21 @@ func New(cs modules.ConsensusSet, tpool modules.TransactionPool, w modules.Walle
 // Close terminates all ongoing processes involving the miner, enabling garbage
 // collection.
 func (m *Miner) Close() error {
+	if err := m.tg.Stop(); err != nil {
+		return err
+	}
+
 	m.mu.Lock()
 	defer m.mu.Unlock()
```

Comment:
@VoidingWarranties for future reference, `saveSync` should be called with a write lock. The persist file should be treated as an extension of the object's state, which means you need a write lock when writing to it. This makes more sense if you frame it in terms of "only one goroutine should write to the file at a time."